src/ir/auth_logic/ast_string.h

```diff
@@ -48,58 +50,116 @@ std::string ToString(AstNodeVariantType Node) {
           },
           [](ConditionalAssertion conditional_assertion) {
             std::string lhs = ToString(conditional_assertion.lhs());
             std::vector<std::string> rhs;
             for (const BaseFact& base_fact : conditional_assertion.rhs()) {
               rhs.push_back(ToString(base_fact));
             }
-            return absl::StrJoin({lhs, absl::StrJoin(rhs, ",")}, " :- ");
+            return absl::StrJoin({lhs, absl::StrJoin(rhs, ", ")}, " :- ");
           },
           [](Assertion assertion) {
             return std::visit(
                 raksha::utils::overloaded{
-                    [](Fact fact) { return ToString(fact); },
+                    [](Fact fact) { return absl::StrCat(ToString(fact), "."); },
                     [](ConditionalAssertion conditional_assertion) {
-                      return ToString(conditional_assertion);
+                      return absl::StrCat(ToString(conditional_assertion), ".");
                     }},
                 assertion.GetValue());
           },
-          [](datalog::Argument argument) {
-            return absl::StrJoin(
-                {argument.argument_name(), argument.argument_type().name()},
-                " : ");
+          [](SaysAssertion saysAssertion) {
+            std::vector<std::string> assertion_strings;
+            assertion_strings.reserve(saysAssertion.assertions().size());
+            for (const Assertion& assertion : saysAssertion.assertions()) {
+              assertion_strings.push_back(ToString(assertion));
+            }
+            return absl::StrCat(ToString(saysAssertion.principal()),
+                                " says {\n",
+                                absl::StrJoin(assertion_strings, "\n"), "}");
+          },
+          [](Query query) {
+            return absl::StrCat("Query(", query.name(), " = ",
+                                ToString(query.principal()), " says ",
+                                ToString(query.fact()), "?)");
+          },
+          [](Program program) {
+            std::vector<std::string> relation_decl_strings;
+            relation_decl_strings.reserve(
+                program.relation_declarations().size());
+            for (const datalog::RelationDeclaration& rel_decl :
+                 program.relation_declarations()) {
+              relation_decl_strings.push_back(ToString(rel_decl));
+            }
+            std::vector<std::string> says_assertion_strings;
+            says_assertion_strings.reserve(program.says_assertions().size());
+            for (const SaysAssertion& says_assertion :
+                 program.says_assertions()) {
+              says_assertion_strings.push_back(ToString(says_assertion));
+            }
+            std::vector<std::string> query_strings;
+            query_strings.reserve(program.queries().size());
+            for (const Query& query : program.queries()) {
+              query_strings.push_back(ToString(query));
+            }
```

Comment:
We can replace these loops with `utils::MapIter`?

Reply:
fixed